src/services/document-import-task-processor.js

```diff
@@ -30,10 +39,43 @@ export class DocumentImportTaskProcessor {
       throw new Error('Cancellation requested');
     }
 
-    await Promise.all(response.users.map(user => this.userService.ensureExternalUser({
+    await Promise.all(documentExport.users.map(user => this.userService.ensureExternalUser({
       _id: user._id,
       username: user.username,
       hostName: batchParams.hostName
     })));
+
+    const sortedRevisions = documentExport.revisions.sort(by(revision => revision.order));
+
+    await this.transactionRunner.run(async session => {
+      for (const revision of sortedRevisions) {
+        const previousRevision = await this.documentService.getCurrentDocumentRevisionByKey(revision.key);
+
+        const user = await this.userService.getUserById(revision.createdBy);
+        const baseUrl = getImportSourceBaseUrl(importSource);
+        const docUrl = getDocUrl(revision.key);
+
+        for (const resource of revision.cdnResources) {
+          const url = `${documentExport.cdnRootUrl}/${resource}`;
+          await this.cdn.uploadObjectFromUrl(resource, url);
+        }
+
+        const mappedRevision = {
+          ...revision,
+          origin: `${DOCUMENT_ORIGIN.external}/${batchParams.hostName}`,
+          originUrl: `${baseUrl}${docUrl}`,
+          appendTo: previousRevision
+            ? {
+              key: revision.key,
+              ancestorId: previousRevision._id
+            }
+            : null
+        };
+
+        await this.documentService.copyDocumentRevision({ doc: mappedRevision, user, databaseSession: session });
```

Comment:
We could consider a cancellation after this step. You basically created a full new revision.

Reply:
Cancelling here would make this into partial import that affects the user.
The import will not have been finalised but the user will already see new document revisions.
This is the reason why I've wrapped this iteration in a transaction. And if the graceful shutdown period would not give it enough time to finish the loop, then we would not persist a partial state and fully resume processing.

Reply:
I know the user would see a partial import, mainly I'm thinking that we saw documents with 100+ revisions and that could take a while to process. If you import 50 of them now and mark the import as failed, the user can still see that it was failed and resume importing from where it left off.

Reply:
By user here you mean just the admin. I mean all users, which is the most common case, the people that see the documents. I would not go for a partial state that was not specifically intended by the importer 🤷‍♀️ 
The case we are talking about is an edge case either way, so I'd treat it as one and opt for all-or-nothing.

Reply:
Sorry to swoop in after this was already resolved, but I think we could find a secure middle ground here that also caters better to documents with a huge amount of revisions or files: We can split the whole document import process into three stages:

1. Ensure all users are created
    Doesn't take long, cancellation afterwards (or even in-between) is no problem, they will be just re-imported on the next attempt

2. Copy all CDN files
    Can potentially take very long, has to be interruptible! But cancellation in-between is no problem, they will be just re-imported on the next attempt, so we should ask if cancellation is requested after each individual file copy process

3. Insert all revisions into the db and create the new "latest" document. This has to be transactional and with a document lock, cancellation of this task in-between and afterwards must not be possible anymore.

The good news is: Until we reach stage 3 everything else is kind of idempotent (if you consider overwriting an existing file as idempotent).

So in my opinion, the code above should be rearranged regarding 2 aspects:
* First we import all the files (with option to cancel after each), then we do the final db transaction
* Check that the revision boundary between the revision returned by `getCurrentDocumentRevisionByKey` is the correct ancestor to the first newly imported revision (there could have been an import in between and they would not match anymore)

Reply:
Oh this sounds good, basically taking the CDN copy out of the transaction as it is repeatable and has no impact on the user. And keeping the revisions as a transaction (plus the doc lock, will add that).

Will make the changes, thanks!

Reply:
Made the changes as you suggested, please double check 🙏

Reply:
Discovered a conflict with the lock, I will fix that first and the ask you to check again.

Reply:
Done, ready to be checked again.